### PR TITLE
KAFKA-16558 Implemented HeartbeatRequestState toStringBase() and added a test for it

### PR DESCRIPTION
**Pull request by brenden20**
I implemented HeartbeatRequestState.toStringBase(), I also added 2 getter methods for testing purposes. I added a test for the new method, which is passing. Test may/may not be necessary, added just in case.

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 1]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
         }
     }
 
+    @Test
+    public void testHeartBeatRequestStateToStringBase() {
+        final long retryBackoffMs = 100;
+        final long retryBackoffMaxMs = 1000;
+        LogContext logContext = new LogContext();
+        HeartbeatRequestState heartbeatRequestState1 = new HeartbeatRequestState(
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Super nit: we can drop the `1` at the end of the variable name, right?

```suggestion
        HeartbeatRequestState heartbeatRequestState = new HeartbeatRequestState(
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 2]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
...
+                time,
+                10,
+                retryBackoffMs,
+                retryBackoffMaxMs,
+                .2
+        );
+
+        RequestState requestState = new RequestState(
+                logContext,
+                "org.apache.kafka.clients.consumer.internals.HeartbeatRequestManager$HeartbeatRequestState",
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Perhaps we could make `HeartbeatRequestState` package-protected, then we could do this:

```suggestion
                HeartbeatRequestManager.HeartbeatRequestState.class.getName(),
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 3]
```diff
@@ -478,6 +478,23 @@ public void resetTimer() {
...
+        @Override
+        public String toStringBase() {
+            return super.toStringBase() +
+                    ", heartbeatTimer=" + heartbeatTimer +
+                    ", heartbeatIntervalMs=" + heartbeatIntervalMs;
+        }
+
+        // Visible for testing
+        protected Timer heartbeatTimer() {
+            return this.heartbeatTimer;
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]:
```suggestion
            return heartbeatTimer;
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 4]
```diff
@@ -478,6 +478,23 @@ public void resetTimer() {
...
+        }
+
+        // Visible for testing
+        protected long heartbeatIntervalMs() {
+            return this.heartbeatIntervalMs;
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]:
```suggestion
            return heartbeatIntervalMs;
```

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks @brenden20! This looks good. Minor nits only.

**Comment by kirktrue** [comment]
The test failures are unrelated, FWIW.

**Comment by brenden20** [comment]
@kirktrue thank you for the suggestions, I have implemented and pushed your suggestions. Let me know if everything looks good!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 5]
```diff
@@ -452,7 +452,7 @@ private void handleFatalFailure(Throwable error) {
      * object extends {@link RequestState} to enable exponential backoff and duplicated request handling. The two fields
      * that it holds are:
      */
-    static class HeartbeatRequestState extends RequestState {
+    protected static class HeartbeatRequestState extends RequestState {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:456]: this is for testing purpose right? add a // Visible for testing comment

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 6]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
...
 
+    @Test
+    public void testHeartBeatRequestStateToStringBase() {
+        final long retryBackoffMs = 100;
+        final long retryBackoffMaxMs = 1000;
+        LogContext logContext = new LogContext();
+        HeartbeatRequestState heartbeatRequestState = new HeartbeatRequestState(
+                logContext,
+                time,
+                10,
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: extract this out as a variable for testing later

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 7]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
...
+
+        RequestState requestState = new RequestState(
+                logContext,
+                HeartbeatRequestManager.HeartbeatRequestState.class.getName(),
+                retryBackoffMs,
+                retryBackoffMaxMs
+        );
+
+        String target = requestState.toStringBase() +
+                ", heartbeatTimer=" + heartbeatRequestState.heartbeatTimer() +
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:178]: can we just apply `time.timer(heartbeatIntervalMs)`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 8]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
...
+                ", heartbeatIntervalMs=" + heartbeatRequestState.heartbeatIntervalMs();
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Just test against `heartbeatIntervalMs`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 9]
```diff
@@ -483,6 +483,23 @@ public void resetTimer() {
...
         }
 
+        @Override
+        public String toStringBase() {
+            return super.toStringBase() +
+                    ", heartbeatTimer=" + heartbeatTimer +
+                    ", heartbeatIntervalMs=" + heartbeatIntervalMs;
+        }
+
+        // Visible for testing
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:494]: i don't think we need the `heartbeatTimer()` and `heartbeatIntervalMs()` here.  see the comment below.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 10]
```diff
@@ -152,6 +152,34 @@ public void cleanup() {
         }
     }
 
+    @Test
+    public void testHeartBeatRequestStateToStringBase() {
+        final long retryBackoffMs = 100;
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: i think we should try to consistently apply final to all variables if we want to use final.

**Comment by philipnee** [comment]
hi @brenden20 - left some comments. lmk what do you think.

**Review by kirktrue — APPROVED** [review]
LGTM!

**Comment by kirktrue** [comment]
@brenden20—can we mark this _Ready for review_?

**Comment by brenden20** [comment]
Marked ready to review @kirktrue

**Review by philipnee — APPROVED** [review]
lgtm thanks for the patch

**Comment by brenden20** [comment]
Merge issues, moved PR to https://github.com/apache/kafka/pull/16373 and closing this one

**Pull request closed without merging**